File: LHDS.AdminPortal.Api.Tests.Acceptance/Brokers/ApiBroker.Suppliers.cs
// ---------------------------------------------------------------
// Copyright (c) North East London ICB. All rights reserved.
// ---------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using LHDS.AdminPortal.Api.Tests.Acceptance.Models.OdataResponses;
using LHDS.AdminPortal.Api.Tests.Acceptance.Models.Suppliers;

namespace LHDS.AdminPortal.Api.Tests.Acceptance.Brokers
{
    public partial class ApiBroker
    {
        private const string suppliersRelativeUrl = "api/suppliers";
        private const string suppliersRelativeOdataUrl = "odata/suppliers";

        public async ValueTask<Supplier> PostSupplierAsync(Supplier supplier) =>
            await this.apiFactoryClient.PostContentAsync(suppliersRelativeUrl, supplier);

        public async ValueTask<Supplier> GetSupplierByIdAsync(Guid supplierId) =>
            await this.apiFactoryClient.GetContentAsync<Supplier>($"{suppliersRelativeUrl}/{supplierId}");

        public async ValueTask<List<Supplier>> GetAllSuppliersAsync()
        {
            OdataResponse<Supplier> response =
                await this.apiFactoryClient.GetContentAsync<OdataResponse<Supplier>>($"{suppliersRelativeOdataUrl}/");

            return response.Items;
        }

        public async ValueTask<List<Supplier>> FilterSuppliersAsync(string supplierName)
        {
            OdataResponse<Supplier> response = await this.apiFactoryClient.GetContentAsync<OdataResponse<Supplier>>(
                $"{suppliersRelativeOdataUrl}/?$filter=name eq '{supplierName}'");

            return response.Items;
        }

        public async ValueTask<List<Supplier>> GetAllSuppliersOrderedDescendingAsync()
        {
            OdataResponse<Supplier> response = await this.apiFactoryClient.GetContentAsync<OdataResponse<Supplier>>(
                $"{suppliersRelativeOdataUrl}/?$orderby=createddate desc");

            return response.Items;
        }

        public async ValueTask<List<Supplier>> GetAllSupplierIngestionTrackingExpandsAsync()
        {
            OdataResponse<Supplier> response = await this.apiFactoryClient.GetContentAsync<OdataResponse<Supplier>>(
                $"{suppliersRelativeOdataUrl}/" +
                    $"?$expand=ingestiontrackings($orderby=CreatedDate asc)&$orderby=createddate asc");

            return response.Items;
        }

        public async ValueTask<List<Supplier>> FindSupplierByIdAsync(Guid supplierId) =>
          await this.apiFactoryClient.GetContentAsync<List<Supplier>>(
              $"{suppliersRelativeUrl}/?$filter=Id eq {supplierId}");

        public async ValueTask<Supplier> PutSupplierAsync(Supplier supplier) =>
            await this.apiFactoryClient.PutContentAsync(suppliersRelativeUrl, supplier);

        public async ValueTask<Supplier> DeleteSupplierByIdAsync(Guid supplierId) =>
            await this.apiFactoryClient.DeleteContentAsync<Supplier>($"{suppliersRelativeUrl}/{supplierId}");
    }
}

File: LHDS.AdminPortal.Configurations/lhds.adminportal.configurations.client/src/brokers/apiBroker.suppliers.ts
import { Supplier } from "../models/suppliers/supplier";
import ApiBroker from "./apiBroker";

class SupplierBroker {
    relativeSupplierUrl = '/api/suppliers';

    private apiBroker: ApiBroker = new ApiBroker();

    async PostSupplierAsync(supplier: Supplier) {
        return await this.apiBroker.PostAsync(this.relativeSupplierUrl, supplier)
            .then(result => new Supplier(result.data));
    }

    async GetAllSuppliersAsync(queryString: string) {
        const url = this.relativeSupplierUrl + queryString;

        return await this.apiBroker.GetAsync(url)
            .then(result => result.data.map((supplier: Supplier) => new Supplier(supplier)));
    }

    async GetSupplierByIdAsync(id: string) {
        const url = `${this.relativeSupplierUrl}/${id}`;

        return await this.apiBroker.GetAsync(url)
            .then(result => new Supplier(result.data));
    }

    async PutSupplierAsync(supplier: Supplier) {
        return await this.apiBroker.PutAsync(this.relativeSupplierUrl, supplier)
            .then(result => new Supplier(result.data));
    }

    async DeleteSupplierByIdAsync(id: string) {
        const url = `${this.relativeSupplierUrl}/${id}`;

        return await this.apiBroker.DeleteAsync(url)
            .then(result => new Supplier(result.data));
    }
}
export default SupplierBroker;